core/src/worker/workflow/machines/workflow_machines.rs

```diff
@@ -598,6 +599,7 @@ impl WorkflowMachines {
             }
         }
 
+        let mut update_admitted_event_messages = HashMap::<String, IncomingProtocolMessage>::new();
```

Comment:
Since this is just on the stack you might actually be able to use `&str` as the keys and save some clones. Not a big deal if it doesn't immediately work.

Reply:
That runs into lifetime violations below, where we create `msg` in a block but want to obtain the key from it:

```rust
            if matches!(
                event.attributes,
                Some(history_event::Attributes::WorkflowExecutionUpdateAdmittedEventAttributes(_)),
            ) {
                let msg = IncomingProtocolMessage::try_from(&event).context(
                    "Failed to create protocol message from WorkflowExecutionUpdateAdmittedEvent",
                )?;
                if self.replaying {
                    update_admitted_event_messages.insert(msg.protocol_instance_id.clone(), msg);
```

Reply:
Ya in principle you could put the binding for the `msg` outside of that block (ie: same level as the map) but keep assigning it there, and then the lifetime would be equal to the map.

But, honestly, not important lol. Probably more confusing to read than it's worth.